### PR TITLE
[VL] Followup for native write files

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

This pr is a followup for https://github.com/oap-project/gluten/pull/3690 to solve some issues:

- Remove `HiveFileFormat` for Spark3.4 shim
- Add an object mapper `VeloxWriteFilesMetrics` for velox table write result json string
- Add task output metrics
- Add number of dynamic part metric
- Use thread local to propagate and clear the writePath
- Make `VeloxColumnarWriteFilesExec` inherit GlutenPlan to avoid wrong fallback report

## How was this patch tested?

add test

### CODE REVIEW

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala [hunk 1]
```diff
@@ -58,81 +96,77 @@ class VeloxColumnarWriteFilesExec(
...
             }
           }
-        }
 
-        // TODO: need to get the partition Internal row?
-        val stats = BasicWriteTaskStats(Seq.empty, loadedCb.numRows() - 1, numBytes, numWrittenRows)
-        val summary =
-          ExecutedWriteSummary(updatedPartitions = updatedPartitions, stats = Seq(stats))
+          // Reports bytesWritten and recordsWritten to the Spark output metrics.
+          Option(TaskContext.get()).map(_.taskMetrics().outputMetrics).foreach {
```
Comment (JkSelf) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala:144]: Why need to update the task metrics?

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala [hunk 2]
```diff
@@ -58,81 +96,77 @@ class VeloxColumnarWriteFilesExec(
...
+              outputMetrics.setRecordsWritten(numWrittenRows)
+          }
 
-        val result = WriteTaskResult(
-          new TaskCommitMessage(addedAbsPathFiles.toMap -> updatedPartitions),
-          summary)
-        Iterator.single(result)
-    }
-  }
+          val partitionsInternalRows = updatedPartitions.map {
```
Comment (JkSelf) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala:150]: Can we also pick you mentioned related unit test in this PR to test this change?

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala (hunk 1 above)
Comment (ulysses-you) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala:144]: it is used to show the task/stage metrics in UI, e.g.,
image

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala (hunk 2 above)
Comment (ulysses-you) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala:150]: yea, the test `numParts` is for this code path

**Comment by ulysses-you** [comment]
@JkSelf any more comments ?

**Comment by JkSelf** [comment]
> @JkSelf any more comments ?

LGTM. The ck backend failed. Can you help to fix?

**Comment by ulysses-you** [comment]
it seems it's irrelevant with this pr, all tests paased but finally jenkin failed to post status..

**Review by JkSelf — APPROVED** [review]
LGTM.

**Pull request merged**